File: packages/python/onnx9000-core/src/onnx9000/core/models/efficientnet.py
"""Efficientnet."""

from typing import Any

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.ops import add, flatten, global_average_pool, mul, reshape
from onnx9000.core.primitives import (
    BatchNormalization,
    ConvND,
    DepthwiseConv,
    Gemm,
    Sigmoid,
    Silu,
)


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:  # noqa: D103
    """Get param."""
    return Variable(name=name, shape=shape, dtype=dtype)


class SqueezeExcitation:
    """Squeeze-and-Excitation block."""

    def __init__(self, in_channels: int, squeeze_channels: int, prefix: str = ""):  # noqa: D107
        """Init."""
        self.prefix = prefix
        self.in_channels = in_channels
        self.squeeze_channels = squeeze_channels
        self.fc1 = Gemm(trans_b=1)
        self.act = Silu()
        self.fc2 = Gemm(trans_b=1)
        self.scale_act = Sigmoid()

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        scale = global_average_pool(x)
        scale = flatten(scale)

        fc1_w = get_param(f"{self.prefix}.fc1.weight", [self.squeeze_channels, self.in_channels])
        fc1_b = get_param(f"{self.prefix}.fc1.bias", [self.squeeze_channels])
        scale = self.fc1(scale, fc1_w, fc1_b)
        scale = self.act(scale)

        fc2_w = get_param(f"{self.prefix}.fc2.weight", [self.in_channels, self.squeeze_channels])
        fc2_b = get_param(f"{self.prefix}.fc2.bias", [self.in_channels])
        scale = self.fc2(scale, fc2_w, fc2_b)
        scale = self.scale_act(scale)

        # Reshape to [N, C, 1, 1] to broadcast with x
        # The input x shape is [N, C, H, W]
        # In ONNX, a reshape needs a shape tensor
        from onnx9000.core.ops import constant

        shape_tensor = constant([-1, self.in_channels, 1, 1], dtype=7)  # int64
        scale = reshape(scale, shape_tensor)

        return mul(x, scale)


class MBConv:
    """Mobile Inverted Bottleneck Convolution."""

    def __init__(  # noqa: D107
        self,
        in_channels: int,
        out_channels: int,
        expand_ratio: int,
        stride: int,
        kernel_size: int,
        prefix: str = "",
    ):
        """Init."""
        self.prefix = prefix
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.stride = stride
        self.use_res_connect = self.stride == 1 and in_channels == out_channels

        hidden_dim = in_channels * expand_ratio
        self.expand_ratio = expand_ratio

        if expand_ratio != 1:
            self.expand_conv = ConvND(2, in_channels, hidden_dim, kernel_size=1, bias=False)
            self.bn0 = BatchNormalization(hidden_dim)
            self.act0 = Silu()

        self.depthwise_conv = DepthwiseConv(
            2,
            hidden_dim,
            kernel_size=kernel_size,
            stride=stride,
            padding=kernel_size // 2,
            bias=False,
        )
        self.bn1 = BatchNormalization(hidden_dim)
        self.act1 = Silu()

        self.se = SqueezeExcitation(hidden_dim, max(1, in_channels // 4), prefix=f"{prefix}.se")

        self.project_conv = ConvND(2, hidden_dim, out_channels, kernel_size=1, bias=False)
        self.bn2 = BatchNormalization(out_channels)

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        identity = x

        if self.expand_ratio != 1:
            x = self.expand_conv(
                x,
                get_param(
                    f"{self.prefix}.expand_conv.weight",
                    [self.expand_conv.out_channels, self.expand_conv.in_channels, 1, 1],
                ),
            )
            x = self.bn0(
                x,
                get_param(f"{self.prefix}.bn0.weight", [self.bn0.num_features]),
                get_param(f"{self.prefix}.bn0.bias", [self.bn0.num_features]),
                get_param(f"{self.prefix}.bn0.running_mean", [self.bn0.num_features]),
                get_param(f"{self.prefix}.bn0.running_var", [self.bn0.num_features]),
            )
            x = self.act0(x)

        x = self.depthwise_conv(
            x,
            get_param(
                f"{self.prefix}.depthwise_conv.weight",
                [
                    self.depthwise_conv.out_channels,
                    1,
                    self.depthwise_conv.kernel_size,
                    self.depthwise_conv.kernel_size,
                ],
            ),
        )
        x = self.bn1(
            x,
            get_param(f"{self.prefix}.bn1.weight", [self.bn1.num_features]),
            get_param(f"{self.prefix}.bn1.bias", [self.bn1.num_features]),
            get_param(f"{self.prefix}.bn1.running_mean", [self.bn1.num_features]),
            get_param(f"{self.prefix}.bn1.running_var", [self.bn1.num_features]),
        )
        x = self.act1(x)

        x = self.se(x)

        x = self.project_conv(
            x,
            get_param(
                f"{self.prefix}.project_conv.weight",
                [self.project_conv.out_channels, self.project_conv.in_channels, 1, 1],
            ),
        )
        x = self.bn2(
            x,
            get_param(f"{self.prefix}.bn2.weight", [self.bn2.num_features]),
            get_param(f"{self.prefix}.bn2.bias", [self.bn2.num_features]),
            get_param(f"{self.prefix}.bn2.running_mean", [self.bn2.num_features]),
            get_param(f"{self.prefix}.bn2.running_var", [self.bn2.num_features]),
        )

        if self.use_res_connect:
            x = add(identity, x)

        return x


class EfficientNet:
    """EfficientNet implementation built using IR macros/primitives."""

    def __init__(self, num_classes: int = 1000):  # noqa: D107
        """Init."""
        self.num_classes = num_classes
        self.stem_conv = ConvND(2, 3, 32, kernel_size=3, stride=2, padding=1, bias=False)
        self.stem_bn = BatchNormalization(32)
        self.stem_act = Silu()

        self.block1 = MBConv(32, 16, expand_ratio=1, stride=1, kernel_size=3, prefix="block1")
        self.block2 = MBConv(16, 24, expand_ratio=6, stride=2, kernel_size=3, prefix="block2")

        self.head_conv = ConvND(2, 24, 1280, kernel_size=1, bias=False)
        self.head_bn = BatchNormalization(1280)
        self.head_act = Silu()

        self.classifier = Gemm(trans_b=1)

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        x = self.stem_conv(
            x,
            get_param(
                "stem_conv.weight", [self.stem_conv.out_channels, self.stem_conv.in_channels, 3, 3]
            ),
        )
        x = self.stem_bn(
            x,
            get_param("stem_bn.weight", [self.stem_bn.num_features]),
            get_param("stem_bn.bias", [self.stem_bn.num_features]),
            get_param("stem_bn.running_mean", [self.stem_bn.num_features]),
            get_param("stem_bn.running_var", [self.stem_bn.num_features]),
        )
        x = self.stem_act(x)

        x = self.block1(x)
        x = self.block2(x)

        x = self.head_conv(
            x,
            get_param(
                "head_conv.weight", [self.head_conv.out_channels, self.head_conv.in_channels, 1, 1]
            ),
        )
        x = self.head_bn(
            x,
            get_param("head_bn.weight", [self.head_bn.num_features]),
            get_param("head_bn.bias", [self.head_bn.num_features]),
            get_param("head_bn.running_mean", [self.head_bn.num_features]),
            get_param("head_bn.running_var", [self.head_bn.num_features]),
        )
        x = self.head_act(x)

        x = global_average_pool(x)
        x = flatten(x)
        x = self.classifier(
            x,
            get_param("classifier.weight", [self.num_classes, 1280]),
            get_param("classifier.bias", [self.num_classes]),
        )

        return x


def efficientnet_b0(**kwargs: Any) -> EfficientNet:  # noqa: D103
    """Efficientnet b0."""
    return EfficientNet(**kwargs)

File: packages/js/core/src/models/efficientnet.ts
import { Tensor } from '../ir/tensor.js';
import { BatchNormalization, ConvND, DepthwiseConv, Gemm, Sigmoid, Silu } from '../primitives.js';

function getParam(
  name: string,
  shape: number[],
  dtype: ReturnType<typeof JSON.parse> = 'float32',
): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: ReturnType<typeof JSON.parse>): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class SqueezeExcitation {
  public prefix: string;
  public inChannels: number;
  public squeezeChannels: number;
  public fc1: Gemm;
  public act: Silu;
  public fc2: Gemm;
  public scaleAct: Sigmoid;

  constructor(inChannels: number, squeezeChannels: number, prefix: string = '') {
    this.prefix = prefix;
    this.inChannels = inChannels;
    this.squeezeChannels = squeezeChannels;
    this.fc1 = new Gemm(1.0, 1.0, 0, 1);
    this.act = new Silu();
    this.fc2 = new Gemm(1.0, 1.0, 0, 1);
    this.scaleAct = new Sigmoid();
  }

  call(x: Tensor): Tensor {
    let scale = recordOp('GlobalAveragePool', [x]);
    scale = recordOp('Flatten', [scale]);

    const fc1W = getParam(`${this.prefix}.fc1.weight`, [this.squeezeChannels, this.inChannels]);
    const fc1B = getParam(`${this.prefix}.fc1.bias`, [this.squeezeChannels]);
    scale = this.fc1.call(scale, fc1W, fc1B);
    scale = this.act.call(scale);

    const fc2W = getParam(`${this.prefix}.fc2.weight`, [this.inChannels, this.squeezeChannels]);
    const fc2B = getParam(`${this.prefix}.fc2.bias`, [this.inChannels]);
    scale = this.fc2.call(scale, fc2W, fc2B);
    scale = this.scaleAct.call(scale);

    const shapeTensor = recordOp('Constant', [], { value: [-1, this.inChannels, 1, 1] });
    scale = recordOp('Reshape', [scale, shapeTensor]);

    return recordOp('Mul', [x, scale]);
  }
}

export class MBConv {
  public prefix: string;
  public inChannels: number;
  public outChannels: number;
  public expandRatio: number;
  public stride: number;
  public useResConnect: boolean;

  public expandConv?: ConvND;
  public bn0?: BatchNormalization;
  public act0?: Silu;

  public depthwiseConv: DepthwiseConv;
  public bn1: BatchNormalization;
  public act1: Silu;
  public se: SqueezeExcitation;
  public projectConv: ConvND;
  public bn2: BatchNormalization;

  constructor(
    inChannels: number,
    outChannels: number,
    expandRatio: number,
    stride: number,
    kernelSize: number,
    prefix: string = '',
  ) {
    this.prefix = prefix;
    this.inChannels = inChannels;
    this.outChannels = outChannels;
    this.stride = stride;
    this.useResConnect = this.stride === 1 && inChannels === outChannels;
    this.expandRatio = expandRatio;

    const hiddenDim = inChannels * expandRatio;

    if (expandRatio !== 1) {
      this.expandConv = new ConvND(2, inChannels, hiddenDim, 1, 1, 0, 1, 1, false);
      this.bn0 = new BatchNormalization(hiddenDim);
      this.act0 = new Silu();
    }

    this.depthwiseConv = new DepthwiseConv(
      2,
      hiddenDim,
      kernelSize,
      stride,
      Math.floor(kernelSize / 2),
      1,
      false,
    );
    this.bn1 = new BatchNormalization(hiddenDim);
    this.act1 = new Silu();

    this.se = new SqueezeExcitation(
      hiddenDim,
      Math.max(1, Math.floor(inChannels / 4)),
      `${prefix}.se`,
    );

    this.projectConv = new ConvND(2, hiddenDim, outChannels, 1, 1, 0, 1, 1, false);
    this.bn2 = new BatchNormalization(outChannels);
  }

  call(x: Tensor): Tensor {
    let identity = x;

    if (this.expandRatio !== 1 && this.expandConv && this.bn0 && this.act0) {
      x = this.expandConv.call(
        x,
        getParam(`${this.prefix}.expand_conv.weight`, [
          this.expandConv.outChannels,
          this.expandConv.inChannels,
          1,
          1,
        ]),
      );
      x = this.bn0.call(
        x,
        getParam(`${this.prefix}.bn0.weight`, [this.bn0.numFeatures]),
        getParam(`${this.prefix}.bn0.bias`, [this.bn0.numFeatures]),
        getParam(`${this.prefix}.bn0.running_mean`, [this.bn0.numFeatures]),
        getParam(`${this.prefix}.bn0.running_var`, [this.bn0.numFeatures]),
      );
      x = this.act0.call(x);
    }

    const ks = Array.isArray(this.depthwiseConv.kernelSize)
      ? this.depthwiseConv.kernelSize[0]
      : this.depthwiseConv.kernelSize;
    x = this.depthwiseConv.call(
      x,
      getParam(`${this.prefix}.depthwise_conv.weight`, [
        this.depthwiseConv.outChannels!,
        1,
        ks!,
        ks!,
      ]),
    );
    x = this.bn1.call(
      x,
      getParam(`${this.prefix}.bn1.weight`, [this.bn1.numFeatures]),
      getParam(`${this.prefix}.bn1.bias`, [this.bn1.numFeatures]),
      getParam(`${this.prefix}.bn1.running_mean`, [this.bn1.numFeatures]),
      getParam(`${this.prefix}.bn1.running_var`, [this.bn1.numFeatures]),
    );
    x = this.act1.call(x);

    x = this.se.call(x);

    x = this.projectConv.call(
      x,
      getParam(`${this.prefix}.project_conv.weight`, [
        this.projectConv.outChannels,
        this.projectConv.inChannels,
        1,
        1,
      ]),
    );
    x = this.bn2.call(
      x,
      getParam(`${this.prefix}.bn2.weight`, [this.bn2.numFeatures]),
      getParam(`${this.prefix}.bn2.bias`, [this.bn2.numFeatures]),
      getParam(`${this.prefix}.bn2.running_mean`, [this.bn2.numFeatures]),
      getParam(`${this.prefix}.bn2.running_var`, [this.bn2.numFeatures]),
    );

    if (this.useResConnect) {
      x = recordOp('Add', [identity, x]);
    }

    return x;
  }
}

export class EfficientNet {
  public numClasses: number;
  public stemConv: ConvND;
  public stemBn: BatchNormalization;
  public stemAct: Silu;
  public block1: MBConv;
  public block2: MBConv;
  public headConv: ConvND;
  public headBn: BatchNormalization;
  public headAct: Silu;
  public classifier: Gemm;

  constructor(numClasses: number = 1000) {
    this.numClasses = numClasses;
    this.stemConv = new ConvND(2, 3, 32, 3, 2, 1, 1, 1, false);
    this.stemBn = new BatchNormalization(32);
    this.stemAct = new Silu();

    this.block1 = new MBConv(32, 16, 1, 1, 3, 'block1');
    this.block2 = new MBConv(16, 24, 6, 2, 3, 'block2');

    this.headConv = new ConvND(2, 24, 1280, 1, 1, 0, 1, 1, false);
    this.headBn = new BatchNormalization(1280);
    this.headAct = new Silu();

    this.classifier = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor): Tensor {
    x = this.stemConv.call(
      x,
      getParam('stem_conv.weight', [this.stemConv.outChannels, this.stemConv.inChannels, 3, 3]),
    );
    x = this.stemBn.call(
      x,
      getParam('stem_bn.weight', [this.stemBn.numFeatures]),
      getParam('stem_bn.bias', [this.stemBn.numFeatures]),
      getParam('stem_bn.running_mean', [this.stemBn.numFeatures]),
      getParam('stem_bn.running_var', [this.stemBn.numFeatures]),
    );
    x = this.stemAct.call(x);

    x = this.block1.call(x);
    x = this.block2.call(x);

    x = this.headConv.call(
      x,
      getParam('head_conv.weight', [this.headConv.outChannels, this.headConv.inChannels, 1, 1]),
    );
    x = this.headBn.call(
      x,
      getParam('head_bn.weight', [this.headBn.numFeatures]),
      getParam('head_bn.bias', [this.headBn.numFeatures]),
      getParam('head_bn.running_mean', [this.headBn.numFeatures]),
      getParam('head_bn.running_var', [this.headBn.numFeatures]),
    );
    x = this.headAct.call(x);

    x = recordOp('GlobalAveragePool', [x]);
    x = recordOp('Flatten', [x]);
    x = this.classifier.call(
      x,
      getParam('classifier.weight', [this.numClasses, 1280]),
      getParam('classifier.bias', [this.numClasses]),
    );

    return x;
  }
}

export function efficientnetB0(numClasses: number = 1000): EfficientNet {
  return new EfficientNet(numClasses);
}
